man/ffpboot.1.in
.de CW
. nop \s-2\f[C]\\$*\f[]\s+2
..
.de CODE
.in +0.5i
. nop \s-2\f[C]\\$*\f[R]\s+2
.in -0.5i
..
.TH ffpboot 1 "[@]DATE[@]" "Version [@]VERSION[@]" "FFP PHYLOGENY"
.SH NAME
ffpboot \- This program performs bootstrap permutations of the FFP vector.
.SH SYNOPSIS
.BI "ffpboot [" "OPTION" "] ... [" "FILE" "] ..."
.SH DESCRIPTION
.PP
Given no options, the default behavior of the program is to
calculate a bootstrap permutation.
This utility should be used
as a filter before using 
.B ffprwn.
FFPs will be read from standard input if no options are supplied and
.B ffpboot
is called non-interactively (i.e. as part of a pipeline) or with a "-" in the 
argument list.
.SH OPTIONS
.TP
.BI "\-j, --jackknife"
Perform deletion jacknife instead of bootstrap. Default
deletion percentage is 1/e.
.TP
.BI "\-p " "FLOAT" ", --delete-prob=" "FLOAT"	
.RI "Specify jacknife deletion Probabilitiy, " "FLOAT" ", ranging between 0 and 1."
.TP
.BI "\-s " "INT" ", --rand-seed=" "INT"	
.RI "Specify random seed, " "INT" "."
The default is (system time) * (process ID).
.TP
.B  "\-h, --help"
Display help message.
.PP
.SH EXAMPLES
To create bootstrap pseudoreplicate:
.PP
.CODE ffpry -l 6 test*.fna | ffpcol > vector
.CODE ffpboot vector > boot
.PP
To create many pseudoreplicates use the shell's looping
facilities.  This example below creates 100 psuedoreplicates,
using a small snippet of Bash shell scripting -- the equivalent
should be possible using all other shell variants (c-shell, zsh, etc.)
.PP
.CODE for i in $(seq 1 1 100) ; do
.CODE \tffpboot vector > boot.$i
.CODE done
.PP
.RB "The bootstrap pseudoreplicates can be used as input to " "ffprwn" ":"
.PP
.CODE ffpboot vector | ffprwn | ffpjsd
.PP
.RB "To perform a deletion jackknife instead use the " "-j" " option."
.PP
.CODE ffpboot -j -p 0.10 vector > jack
.PP
The above example sets the probability of feature deletion
in the jackknifing process to 10%.
.PP
As of version 3.06, there is a streamlined method of generating
pseudo-replicate trees, which can be used in the phylip program
.CW consense
to build a consensus tree. 
.PP
.CODE for i in {1..100} ; do
.CODE \tffpboot vector | ffprwn | ffpjsd -p species.txt | ffptree -q
.CODE done > intree
.PP
The final output 
.CW intree
contains multiple sets of Newick format tree files which can be 
used directly in 
.CW consense.
.PP
.SH FURTHER DIRECTIONS
Option to create multiple replicates in a single exection.
.SH AUTHOR
This program was written by Gregory E. Sims.
.SH "REPORTING BUGS"
Report bugs to <gesims@lbl.gov>.
.SH COPYRIGHT
Copyright (C) [@]COPY[@] Gregory E. Sims
.br
There is NO WARRANTY, to the extent permitted by law.
.SH "SEE ALSO"
.BR ffpaa(1),
.BR ffpry(1),
.BR ffprwn(1),
.BR ffpjsd(1),
.BR ffpcol(1),
.BR ffpmerge(1),
.BR ffpvprof(1),
.BR ffptree(1),
.BR ffpreprof(1)


NEWS
Version 3.19
	Thanks to Joemar Taggana, some incompatible utilities 
	were discovered in Mac OSX -- including incompatible
	getopt and mktemp.  Corrected confusion in tutorial 
        p. 3 regarding -m (Thanks to Erin Shellman).  Corrected
	spelling of ffpjsd option --hamann to --hamman.

Version 3.18
	Rearranged ungetc character tests to come after
	conversion of pipe to file.  Fixed memory leak in
	hash init().  Added support for large files (> 2GB) on 32
        bit machines.

Version 3.17
	Fixed bug in tempfile creation in function 
        convertPipeToFile().

Version 3.16
	Fixed bug in ffpry with multiple headers parsed as a
	single genome entry.

Version 3.15
	Fixed bug in ffprwn.

Version 3.14
	Added --disable-gui options to ./configure script. This
	disables the perl based gui, so that the Perl/Tk module
	isn't required for installation.

Version 3.13
	Changed row terminal output for all programs from '\t\n' 
	(tab newline) to '\n' (newline) only.  This makes output
	from previous version incompatible with 3.13 but the
	ffps can be easily converted using:

	sed 's/\t$//' ffp > ffp.tmp ; mv ffp.tmp ffp

Version 3.12
	Small bug fix for ffprwn for GNU compiler on 64 bit machines
	where promotion of unsigned int to long unsigned int is
	different.  Thanks to Adrienne Breland at the University
	of Nevada, Reno for finding this.

Version 3.10
	Fixed minor bug in ffpry for fasta records pseudomolecule 
	input ENDING in NNNNNNNN, or other invalid characters.

Version 3.09
	Fixed bug in ffpdf, related to initialized GROUP variable.
	
	Updated license

Version 3.08

	ffpgui

	Added instructions to install a functional version of the
	perl/tk module for using cygwin, in INSTALL.

	Added error messages in ffpgui for running a vocab or 
	relative entropy profile without loading a file first.

	Added additional sections to tutorial.pdf
       


Version 3.07
	Added signal handling to all executables.  
	Broken pipes will cause a termination error in a pipeline
	and will output 'PROGRAM: Broken pipe'

	Expanded ./doc/tutorial.in.
	Updated license.
	Fixed inappropriate default -f setting (changed from 1 to
	2) for ffpvprof.

Version 3.06
	Added ffptree utility, which creates neighbor joining and
	upgma trees.  Date and version automatically updated 
	during build process.
	
        Improved seeding of random generator when no seed is
	provide via command line option:  ffpry, ffpboot, ffpaa

	Added processing via ffptree to ffpgui.  The human 
	readable tree is displayed in the message window.

	Started tutorial file in ./doc directory. 

	Added tutorial.pdf to doc directory

Version 3.05
	Added length test since MAX_WORD_SIZE is 40: ffpry, ffpaa
	ffpre.

Version 3.04
	Added script ffpgui which is a Perl/Tk based GUI wrapper of
	the various utilities.

	Added --keys flag to ffpfilt to force keys to be output
	instead of values only.

Version 3.03
	Fixed overflow error with non-rabin karp hash used in ffpcol.

Version 3.02
	The ffpaat utility has been removed from this package.  The
	purpose of this utility was to count amino acid features in 
	all nine frames from DNA fasta input.  However	obtaining
	a proper translation of gene regions and using ffpaa is 
	generally more informative.  This utility can still be 
	compiled and used from previous versions, but will not longer
	be part of future releases. 

	Updated ffpre to Rabin-Karp hash, also fixed float overflow
	errors and cases of nan, or inf in output.  Text mode is now
	implemented in ffpre.

	Updated ffptxt to Rabin-Karp hash.


Version 3.01
	Implementation of Rabin-Karp hash in ffpaa utility.
	Similar speed gains as in ffpry obtained


	ffpaa -l 20 NC_008253.faa      

	ffpaa (v.3.0)                        ----------- 9.35 s 
	ffpaa (v.3.01)                       ----------- 0.15 s

	Input buffering options removed from both ffpaa and ffpry. The
	optimal buffering size for the system is determined automatically
	by using the stat.st_block size attribute of fstat or the compiler
	default, whichever is greater.  This is an effort to use the
	filesystem's most efficient block size.  
	
	Upcoming v3.02 will include Rabin-Karp implementation of ffpre.

Version 3.0
	Thanks to Nick Loman of University of Birmingham for pointing out
	the cryptic nature of the "key-value" error messages when both
	ffpfilt and ffpcol are used together in a pipeline successively.
	
	All error messages from the utilities state the program issuing
	the error.

	Revamped tmpfile creation for non-seekable piped input, to use
	most efficient buffering.

	Added extra unit tests to confirm piped vs file argument output
	is identical.

	ffpry changes:

	Beginning initial development of Rabin Karp Rolling hash function
	implementations which has replaced the hashing function used
	in version 2 in the ffpry executable.  This has provided a 
	significant speedup in k-mer counting.
	Here is a speed comparison:
	
	ffpry -l 10 -d NC_008253.fna             ------    9.7s (v.3.0)
        ffpry -l 10 -d NC_008253.fna             ------ 1m28.2s (v.2.30

	Removed all-possible/-a flag from ffpry.  The only way to generate
	columnar output from ffpry is via the -f flag and specify a specific
	feature set.

        -f specified features are now reverse complemented and classed as
	is the input sequence by default.	

	Added -r option which disables reverse complementing of the input
	sequence as well as operates on the feature list.  
	
	Fixed ffpry -f long option to correct value as per man file.
	
	Other:

	Fixed segmentation error in ffpboot

	Future directions:

	Rabin-Karp hash function will be implemented in ffpaa, ffptxt and
	ffpre, ffpaat
	
	

Version 2.30
	Fixed bug when encountering IUPAC codes in input files to ffpry.

Version 2.29
	Fixed bug when piping input into ffpmerge.

Version 2.28
	Addition of examples directory with trees from 
	Sims & Kim(2011) PNAS,108,8329-8334

        BUGS: ffpre doesn't report errors for trying to open

src/ffpboot.c
/*****************************************************
* This code is distributed under a Non-commercial use 
* license.  For details see LICENSE.  Use of this
* code must be properly attributed to its author
* Gregory E. Sims provided that its use or derivative 
* use is non-commercial in nature.  Proper attribution        
* can be made by citing:
*
* Sims GE, et al (2009) Alignment-free genome 
* comparison with feature frequency profiles (FFP) and 
* optimal resolutions. Proc. Natl. Acad. Sci. USA.
* 106, 2677-82.
*
* Gregory E. Sims (C) 2010-2012
*
*****************************************************/
#include <stdio.h>
#include <math.h>
#include <getopt.h>
#include <stdlib.h>
#include <ctype.h>
#include <time.h>
#include <errno.h>
#include <limits.h>
#include <unistd.h>
#include <string.h>
#include "vstring.h"
#include "utils.h"
#include "sighandle.h"
#include "../config.h"

char PROG_NAME[] = "ffpboot";
#define COL_SIZE 1000 /**< Initial guess for columns in FFP */
#define DEFAULT_JACK 0.36787944117144232159 /**< Default Jackknife probability of deletion: 1/exp(1) */
#define INIT_ROWSIZE 10

void bootstrap(FILE * fp);
void jacknife(FILE * fp, float);

char usage_str[] = "Usage: %s [OPTIONS] ... [FILE] ...\n\
This program performs bootstrap permutations of the FFP vector\n\n\
Given no options, the defeault behavior of the program is to\n\
calculate a bootstrap permutation.\n\
\t-j, --jackknife\n\
\t-p PROB, --delete-prob=PROB\n\
\t-s INT, --rand-seed=INT\n\
\t-v, --version\n\
\t-h, --help\n\n\
Copyright (c) %s\n\
%s\n\
Contact %s\n";


int main(int argc, char **argv)
{
    FILE *fp;
    int opt;
    char jflag = 0;
    char sflag = 0;
    unsigned svalue = 0;
    float pvalue = DEFAULT_JACK;
    int option_index = 0;

    static struct option long_options[] = {
	{"help", no_argument, 0, 'h'},
	{"delete-prob", required_argument, 0, 'p'},
	{"jackknife", no_argument, 0, 'j'},
	{"rand-seed", required_argument, 0, 's'},
	{"version", no_argument, 0, 's'},
	{0, 0, 0, 0}
    };

    initSignalHandlers();

    while ((opt = getopt_long(argc, argv, "jp:s:vh",
			      long_options, &option_index)) != -1)

	switch (opt) {
	case 'p':
	    pvalue = atof(optarg);
	    break;
	case 'j':
	    jflag = 1;
	    break;
	case 's':
	    sflag = 1;
	    svalue = atoi(optarg);
	    break;
	case 'v':
	    printVersion();
	    exit(EXIT_SUCCESS);
	    break;
	case 'h':
	    printUsageStr();
	    exit(EXIT_SUCCESS);
	    break;
	default:
	    printErrorUsageStr();
	    exit(EXIT_FAILURE);
	    break;
	}

    if (sflag)
	srand((unsigned) svalue);
    else
	srand((unsigned) time(NULL) * getpid());


    if (pvalue > 1 || pvalue < 0) {
	fatal_msg("Jacknife deletion probability must be betwen 0 and 1\n");
    }



    
    argv += optind;

    do {
	fp = stdin;
	if (*argv) {
	    if (!strcmp(*argv, "-"))
		fp = stdin;
	    else if ((fp = fopen(*argv, "r")) == NULL)
	    	fatal_msg("%s: %s\n", *argv, strerror(errno));

	    if ( isDirectory(*argv) ) 
	    	fatal_msg("%s: %s\n",*argv, strerror(EISDIR));

	    argv++;

	} else if (isatty(STDIN_FILENO))
	    printErrorUsageStr();

	if (!isRegularFile(fp)) 
	    fp = convertPipeToFile(fp);

	if (isKeyBased(fp))
	    fatal_msg("Input is not columnar format\n");

	if (jflag)
	    jacknife(fp, pvalue);
	else
	    bootstrap(fp);

	if (fp != stdin)
	    fclose(fp);

    } while (*argv);

    return EXIT_SUCCESS;
}



/**
 *
 * Performs a bootsrap permutation of a columnar FFP.
 *
 * Column numbers are sampled with replacement up to the
 * number of columns in the original matrix and
 * the chosen columns are printed to stdout.
 *
 * @param fp A file pointer to an FFP.
 * @return void
 *
 */



void bootstrap(FILE * fp)
{
    long unsigned i, j;
    unsigned *randCols;
    int rows;
    unsigned *vals;
    unsigned cols;
    unsigned size = COL_SIZE;	/* Inital guess */
    int rowsize = INIT_ROWSIZE;	/* initial guess */
    char s[2];
    int d;


    vals = (unsigned *) malloc(sizeof(unsigned) * size);

    i = 0;
    rows = 0;
    while (!feof(fp)) {
	if (i < size) {
	    d = fscanf(fp, "%u%[\n\r]", &vals[i++], s);
	    if (d == 2) {
		rows++;
		if (rows > rowsize) {
		    rowsize += INIT_ROWSIZE;
		}
	    }
	} else {
	    size += COL_SIZE;
	    vals = (unsigned *) realloc(vals, sizeof(unsigned) * size);
	}
    }

    cols = (unsigned) i / rows;

    //Now we know the number of columns

    randCols = (unsigned *) malloc(sizeof(unsigned) * cols);

    for (i = 0; i < cols; i++)
	randCols[i] = rand() % cols;

    for (i = 0; i < rows; i++) {
	for (j = 0; j < cols-1; j++)
	    printf("%u\t", vals[i * cols + randCols[j]]);
	printf("%u\n", vals[i * cols + randCols[j]]);
    }

    free(vals);
}



/**
 *
 * Performs a jackknife resampling of a columnar FFP.
 *
 * Column numbers are sampled without replacement using
 * a deletion jackknife. Each column has a fixed probability
 * of being deleted equal to the second argument del. 
 * The remainin columns are printed to stdout.
 *
 * @param fp A file pointer to an FFP.
 * @param del The deletion probability.
 * @return void
 *
 */

void jacknife(FILE * fp, float del)
{
    long unsigned i, j;
    char *randCols;
    int rows;
    unsigned *vals;
    unsigned cols;
    unsigned size = COL_SIZE;	/* Inital guess of column number */
    int rowsize = INIT_ROWSIZE;	
    char s[2];
    int d;


    vals = (unsigned *) malloc(sizeof(unsigned) * size);

    i = 0;
    rows = 0;
    while (!feof(fp)) {
	if (i < size) {
	    d = fscanf(fp, "%u%[\n\r]", &vals[i++], s);
	    if (d == 2) {

		rows++;
		if (rows > rowsize) {
		    rowsize += INIT_ROWSIZE;
		}
	    }
	} else {
	    size += COL_SIZE;
	    vals = (unsigned *) realloc(vals, sizeof(unsigned) * size);
	}
    }

    cols = (unsigned) i / rows;

    //Now we know the number of columns

    randCols = (char *) malloc(sizeof(char) * cols);

    for (i = 0; i < cols; i++) {
	randCols[i] = ((float) rand() / INT_MAX > del);
    }

    for (i = 0; i < rows; i++) {
	for (j = 0; j < cols; j++)
	    if (randCols[j])
		printf("%u\t", vals[i * cols + j]);
	printf("%u\n", vals[i * cols + j]);
    }

    free(vals);
}
